### PR TITLE
CI: add codecov to Azure CI

### PR DESCRIPTION
**Pull request by tylerjereddy**
* add a codecov upload to Azure CI (32-bit
Windows)

* I was a bit surprised that we don't
seem to have a `yaml` file for codecov
settings (?), so I didn't add a lint
stage for our codecov config

* reviewers should check carefully that the upload actually happens because the job passed without uploading on my fork (i.e., the upload utility ran, but then exited with `Commit sha does not match Azure build. Please upload with the Codecov repository upload token to resolve issue.
`,  hopefully just because it was on my fork...)

Changes made in this Pull Request:
 - 

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by tylerjereddy** [comment]
I still see `Commit sha does not match Azure build. Please upload with the Codecov repository upload token to resolve issue.` I suppose we could use the token if we really have to. The Azure blog [suggests we don't have to](https://devblogs.microsoft.com/devops/uploading-to-codecov-just-got-easier/), but the example is on Linux so..

**Comment by orbeckst** [comment]
How sensitive is the upload token?  The codecov instructions say

> Add to your codecov.yml
```yaml
codecov:
  token:  xxxxx
```

or Set in your environment variables
```bash
CODECOV_TOKEN="xxxx"
```
On travis we could encrypt the token. Can we add it to Azure encrypted or via configuration?

**Comment by tylerjereddy** [comment]
I've saved the encrypted codecov token in Azure UI configuration, but I think it does not get automatically encrypted. If the CI fails to upload again this time I'll have to use the mapping techniques they describe to de-encrypt [in the docs for Azure secret vars](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#secret-variables)

**Comment by tylerjereddy** [comment]
Ok, I need to decrypt the token still it seems

**Comment by orbeckst** [comment]
Do you have access to the decrypted codecov token, i.e., when you go to https://codecov.io/gh/MDAnalysis/mdanalysis/settings and login with GitHub, can you see it?

**Comment by tylerjereddy** [comment]
Yeah, I just need to follow the Azure docs to use it properly. I'll be slow, but should get it done. Failing that, the new codecov team is pretty responsive to GitHub pings, but will try on my own first.

**Comment by tylerjereddy** [comment]
Now it says `Provided token is not a UUID.` @thomasrockhu Any hints on the Azure + Windows + codecov? I probably have something silly like needing to add quotes around the secret token I pasted in Azure UI or something.

**Comment by thomasrockhu** [comment]
Hmmm, I'm not sure why that is. I copy and pasted my token into this project here (and actually used your yaml as a template): https://dev.azure.com/tomhu1096/codecov-github-test/_build/results?buildId=8&view=logs&j=e4f8c82b-de6b-5575-3950-14744af47318&t=9fea38d8-bc43-58c0-468d-6e85974c7041

Are you adding quotes or spaces or anything like that?

**Comment by thomasrockhu** [comment]
As an example, this is what my settings page looks like in Codecov
image

And this is what I pasted into Azure
image

**Comment by tylerjereddy** [comment]
I checked the "Keep this value secret" option, so I think the decryption component of the workflow may be adding some additional complication.

Is there no reason to keep the codecov tokens secret? If not, then I'll just do away with that and maybe it will work like yours does.

**Comment by thomasrockhu** [comment]
@tylerjereddy apologies, I actually had the `Keep this value secret` checkbox clicked but wanted you to be able to see what the format looked like

**Comment by tylerjereddy** [comment]
My setup should have been the same--I've tried deleting and setting up the variable again with a fresh copy-paste of the token.

**Comment by tylerjereddy** [comment]
Same thing--possible differences with your setup might be:

- yours is a direct push to master branch instead of a PR from a fork
- we are using 32-bit Windows instead of the normal 64-bit used in your example

**Comment by thomasrockhu** [comment]
@tylerjereddy, I think I'm using 32-bit now and it's in a PR. https://dev.azure.com/tomhu1096/codecov-github-test/_build/results?buildId=13&view=logs&jobId=eb841e9f-6e1e-5c29-3ad5-74ec2c658069&j=d420aaaa-0ae7-5c90-2d94-1e61729d0f21&t=2ebca456-40b1-5871-95d6-ad048e69e584

**Comment by thomasrockhu** [comment]
@tylerjereddy, I see this is still an issue. Would you be able to echo `$CODECOV_TOKEN` and see what it's actually trying to spit it out as? (An admin can always and should regenerate that token on the Codecov UI, and leaking this secret has very low impact over a short period of time).

**Comment by tylerjereddy** [comment]
I'll see what I can do--I should maybe check if the token is used anywhere else first, but I think most CI services don't even need it for public repos. The most recent attempt was with quotes added--the first two had no quotes.

**Comment by orbeckst** [comment]
I don't think we use the token elsewhere. Just change it when done.

**Comment by tylerjereddy** [comment]
Ok, I've added an echo for the variable--we'll see what happens I guess

**Comment by thomasrockhu** [comment]
Hmmm... I'm going to fork the repo and make the change to see why this might be happening sometime tonight or tomorrow. Sorry this is blocking your usage of Codecov on AzureCI

**Comment by thomasrockhu** [comment]
@tylerjereddy, we did some work on the Azure pipelines side of things which may have helped with this issue. My build now seems to work, and you can see the build report [here](https://dev.azure.com/mdanalysis/mdanalysis/_build/results?buildId=337&view=logs&j=1cdae27b-58e8-5c64-01e2-519b6cdf4b73&t=7c35adf3-d75f-56e4-8e51-9f2e1c032293). Could you try running this again to see if it's been fixed?

**Comment by tylerjereddy** [comment]
@thomasrockhu Cool, I've rebased and force pushed to trigger the rebuild :)

File: azure-pipelines.yml [hunk 1]
```diff
@@ -55,5 +56,12 @@ jobs:
...
-      pytest .\MDAnalysisTests --disable-pytest-warnings -n 2 -rsx
+      pytest .\MDAnalysisTests --disable-pytest-warnings -n 2 -rsx --cov=MDAnalysis
     displayName: 'Run MDAnalysis Test Suite'
+  - script: |
+      curl -s https://codecov.io/bash | bash
+      echo %CODECOV_TOKEN%
+    condition: succeeded()
+    displayName: 'Codecov Coverage Data Upload'
+    env:
+      CODECOV_TOKEN: $(CODECOV_TOKEN)
```
Comment (thomasrockhu) [inline azure-pipelines.yml:67]:
```suggestion
```

File: azure-pipelines.yml [hunk 2]
```diff
@@ -55,5 +56,12 @@ jobs:
     displayName: 'Build MDAnalysis'
   - powershell: |
       cd testsuite
-      pytest .\MDAnalysisTests --disable-pytest-warnings -n 2 -rsx
+      pytest .\MDAnalysisTests --disable-pytest-warnings -n 2 -rsx --cov=MDAnalysis
     displayName: 'Run MDAnalysis Test Suite'
+  - script: |
+      curl -s https://codecov.io/bash | bash
+      echo %CODECOV_TOKEN%
```
Comment (thomasrockhu) [inline azure-pipelines.yml:63]:
```suggestion
```

**Comment by tylerjereddy** [comment]
thanks, I committed your suggestions just now

**Comment by thomasrockhu** [comment]
@tylerjereddy success!

**Comment by tylerjereddy** [comment]
Great! The PR should be ready to merge when the rest of the CI completes. I assume we can ignore the complaint from `curl` in there since the upload seems to happen anyway.

**Comment by thomasrockhu** [comment]
@tylerjereddy, yes the `curl` complaint is fine (it's actually my fault and I need to find the time to understand why)

**Comment by orbeckst** [comment]
Well, that increased our coverage by almost 1%!

Thank you @thomasrockhu for helping, much appreciated! (And makes me happy to use codecov for other projects!)

**Comment by orbeckst** [comment]
@tylerjereddy I‘ll leave it to you to squash and merge, just in case you want to refine anything.

**Comment by thomasrockhu** [comment]
@orbeckst happy to help!

**Pull request merged**